src/base-xhr.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+var objectMirrors = require('./object-mirrors');
+
+var BaseXHR = function(xhrImpl, onreadystatechange, onprogress, onloadend) {
```

Comment:
they are callbacks for when these event handlers will be set

they will pass the handler that has been set like `(handler) => {}` and then one can do anything with it in the implementation. it is necessary for the shaper case- for the xhrSetup it allows us to throw whenever one would attempt to set these handlers.